### PR TITLE
Data

### PR DESCRIPTION
**Pull request by jordan-singer**

### CODE REVIEW

File: src/Blackbird.py [hunk 1]
```diff
@@ -11,241 +14,283 @@
...
+            `game`: A GameState object which holds the rules of the game
+                BlackBird is intended to learn.
+            `name`: The name of the model.
+            `mctsConfig` : JSON config for MCTS runtime evaluation
+            `networkConfig` : JSON config for creating a new network from NetworkFactory
+            `tensorflowConfig` : Configuaration for tensorflow initialization
+    """
 
-        return wins, draws, losses
+    def __init__(self, game, name, mctsConfig, networkConfig={}, tensorflowConfig={}):
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/Blackbird.py]: The intent was to automatically label saved models with the game they learned.  Is there an easier/more succinct method, do you think?

File: src/DataManager.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        self.Cursor.execute(command, (arch,))
+        self._conn.commit()
+
+    def PutModel(self, archKey, gameType, name, version):
+        command = 'INSERT INTO ModelDim(ArchitectureKey, GameType, Name, Version) VALUES(?,?,?,?);'
+        self.Cursor.execute(command, (archKey, gameType, name, version))
+        self.ModelKey = self.Cursor.lastrowid
+        self.Cursor.execute("""INSERT INTO TrainingStatisticsFact(
+            ModelKey) VALUES (?);""", (self.ModelKey,))
+
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Would you be more comfortable with inserting all aggregated statistics at once?  Or does your hesitance stem from `ModelKey` being in the `TrainingStatisticsFact` table altogether?

File: src/DataManager.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            raise ValueError('statType must be "random", "self", or "MCTS".')
+
+        self.Cursor.execute(command, (*stats.values(), modelKey))
+        self._conn.commit()
+
+    def Close(self):
+        self._conn.close()
+
+    def SetLastModel(self, gameType, name):
+        self.Cursor.execute('SELECT MAX(ModelKey) FROM ModelDim WHERE GameType = ?;', (gameType,))
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Yeah, good find.  I'll update this shortly.

[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Sure, that's a good strategy.  It makes sense that whatever they key is, we shouldn't rely on them being in any particular order, say if we switched to using some autogenerated guid.

File: src/DataManager.py [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+        self._makeSchema(self.Cursor)
+        self.ModelKey = None
+
+    def PutArchitecture(self, arch):
+        command = 'INSERT INTO ArchitectureDim(ArchitectureJSON) VALUES (?);'
+        self.Cursor.execute(command, (arch,))
+        self._conn.commit()
+
+    def PutModel(self, archKey, gameType, name, version):
+        command = 'INSERT INTO ModelDim(ArchitectureKey, GameType, Name, Version) VALUES(?,?,?,?);'
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Are you suggesting that we give some of those function arguments default values?

File: src/Blackbird.py [hunk 5]
```diff
@@ -11,241 +14,283 @@
...
+        state.ParseFromString(serialState)
+        boardDims = np.frombuffer(state.boardDims, dtype=np.int8)
+
+        self.player = state.player,
+        self.mctsEval = state.mctsEval,
+        self.mctsPolicy = np.frombuffer(state.mctsPolicy,
+                                        dtype=np.float),
+        self.board = np.frombuffer(state.boardEncoding,
+                                   dtype=np.int8).reshape(boardDims)
+
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/Blackbird.py]: Agreed.

File: src/DataManager.py [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+    def _makeSchema(self, cursor):
+        check = """SELECT COUNT(name) FROM sqlite_master
+                   WHERE type='table' AND name='TrainingStatisticsFact';"""
+        cursor.execute(check)
+        if cursor.fetchone()[0] == 1:
+            return
+
+        tableStatements = []
+        tableStatements.append("""
+            CREATE TABLE ArchitectureDim(
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Excellent, I'll remove this from the code.

File: src/DataManager.py [hunk 7]
```diff
@@ -0,0 +1,158 @@
...
+
+    def DumpToZip(self, modelKey):
+        self.Cursor.execute('SELECT State FROM GameStateFact WHERE ModelKey = ?;', (modelKey,))
+
+        with gzip.open('data/states_{0}.gz'.format(modelKey), 'wb') as states_out:
+            for state in self.Cursor.fetchall():
+                states_out.write(state[0]+b'\x07\x07\x07')
+
+    def PutTrainingStatistic(self, statType, stats, modelKey):
+        command = """UPDATE TrainingStatisticsFact
```
[the remark this answers is not in the document]
Comment (jordan-singer) [reply src/DataManager.py]: Is there a use case for needing different statistics generated from the same trained model?

[the remark this answers is not in the document]
Comment (Kevdog25) [reply src/DataManager.py]: I was imagining this table being more flexible. 
What if we totally went the other way with the granularity here? If we instead had ~5 columns in it - 
TrainingStatisticsKey, Time, Model1, Model2, WinningModel.

Then, when we stood up the database, we made key 1 random. In my head, at least, that would be a super clean structure that is simple to read and write from and it means that that table does 1 thing.

File: src/DataManager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Kevdog25) [reply src/DataManager.py]: ahh no. Conceptually, if you are writing to ModelDim, you are either updating a model or making a new one. This does not currently support updating.

File: src/DataManager.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Kevdog25) [reply src/DataManager.py]: This is really just about my problem with TrainingStatisticsFact. Not really about this particular line. Although I do think its wonky to do inserts without checks and rely on calling other things in the right order to make it work.

File: src/Blackbird.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kevdog25) [reply src/Blackbird.py]: Could we just use the name of the model? Is this just for better supporting reporting use cases?

**Pull request merged**